File: src/Atlasd/Battlenet/Protocols/Game/Frame.cs
using System;
using System.Collections.Concurrent;

namespace Atlasd.Battlenet.Protocols.Game
{
    class Frame
    {
        public ConcurrentQueue<Message> Messages { get; protected set; }

        public Frame()
        {
            Messages = new ConcurrentQueue<Message>();
        }

        public Frame(ConcurrentQueue<Message> messages)
        {
            Messages = messages;
        }

        public byte[] ToByteArray()
        {
            var framebuf = new byte[0];
            var msgs = new ConcurrentQueue<Message>(Messages); // Clone Messages into local variable

            while (msgs.Count > 0)
            {
                if (!msgs.TryDequeue(out var msg)) break;

                var messagebuf = msg.ToByteArray();
                var buf = new byte[framebuf.Length + messagebuf.Length];

                Buffer.BlockCopy(framebuf, 0, buf, 0, framebuf.Length);
                Buffer.BlockCopy(messagebuf, 0, buf, framebuf.Length, messagebuf.Length);

                framebuf = buf;
            }

            return framebuf;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Other/Frame.vb
Imports System
Imports System.Collections.Concurrent

Namespace AtlasV.Battlenet.Protocols.Game
    Class Frame
        Public Property Messages As ConcurrentQueue(Of Message)

        Public Sub New()
            Messages = New ConcurrentQueue(Of Message)()
        End Sub

        Public Sub New(ByVal varMessages As ConcurrentQueue(Of Message))
            Messages = varMessages
        End Sub

        Public Function ToByteArray(ByVal varProtocolType As ProtocolType) As Byte()
            Dim framebuf = Array.Empty(Of Byte)()
            Dim msgs = New ConcurrentQueue(Of Message)(Messages)
            Dim msg = Nothing

            While Not msgs.IsEmpty '.Count > 0
                If Not msgs.TryDequeue(msg) Then Exit While
                Dim messagebuf = msg.ToByteArray(varProtocolType)
                Dim buf = New Byte(framebuf.Length + messagebuf.Length - 1) {}
                Buffer.BlockCopy(framebuf, 0, buf, 0, framebuf.Length)
                Buffer.BlockCopy(messagebuf, 0, buf, framebuf.Length, messagebuf.Length)
                framebuf = buf
            End While

            Return framebuf
        End Function
    End Class
End Namespace
